std/json.c3

```diff
@@ -0,0 +1,428 @@
+@require_once "std/arithmetic.c3"
+@require_once "std/string.c3"
+@require_once "std/memory.c3"
+@require_once "std/vector.c3"
+@require_once "std/wrappers.c3"
+@require_once "std/iterators.c3"
+
+typedef Cursor : {
+    data      : StringView,
+    position  : isize, // TODO should this be an iterator?
+    allocator : Allocator,  // FIXME use global allocator.
+}
+
+typedef JSON_Node : {
+    type     : int, // 0: null, 1: true, 2: false, 3: string, 4: number,
+                    // 5: array, 6: object
+    value    : Optional<StringView>,
+    children : Optional<Vector<JSON_Node>>,
+}
+
+// TODO something like new<JSON_Node> would be nice, but we don't support
+// function specializations.
+function new_json_node : (node_type : int) -> JSON_Node = {
+    return {
+        node_type,
+        make_optional<StringView>(),
+        make_optional<Vector<JSON_Node>>(),
+    };
+}
+
+function char : (literal : u8[1]&) -> u8 = {
+    return literal[0];
+}
+
+function is_eof : (cursor : Cursor&) -> bool = {
+    return cursor.position >= cursor.data.length;
+}
+
+function peak : (cursor : Cursor&) -> Optional<u8> = {
+    if !cursor:is_eof() {
+        return make_optional(cursor.data.data[cursor.position]);
+    }
+
+    return make_optional<u8>();
+}
+
+function next : (cursor : Cursor&) -> Optional<u8> = {
+    // TODO this should be const (can't use with has_value() right now).
+    let token : Optional<u8> = cursor:peak();
+
+    if token:has_value() {
+        cursor.position += 1;
+    }
+
+    return token;
+}
+
+function is_json_whitespace : (token : u8) -> bool = {
+    // JSON defines whitespace as one of { space, linefeed, carriage return,
+    // horizontal tab }. If we use string literals instead of code points, then
+    // our code doesn't care about the underlying encoding :).
+    return token == char(" ") or token == char("\n")
+           or token == char("\r") or token == char("\t");
```

Comment:
`is_one_of(token, ...)`?

Reply:
Good shout. It optimized badly unfortunately
```asm
0000000000203ce0 <is_json_whitespace__T_u8>:
  203ce0:	b0 01                	mov    $0x1,%al
  203ce2:	40 80 ff 0a          	cmp    $0xa,%dil
  203ce6:	74 09                	je     203cf1 <is_json_whitespace__T_u8+0x11>
  203ce8:	40 0f b6 cf          	movzbl %dil,%ecx
  203cec:	83 f9 20             	cmp    $0x20,%ecx
  203cef:	75 01                	jne    203cf2 <is_json_whitespace__T_u8+0x12>
  203cf1:	c3                   	ret
  203cf2:	40 80 e7 fb          	and    $0xfb,%dil
  203cf6:	40 80 ff 09          	cmp    $0x9,%dil
  203cfa:	0f 94 c0             	sete   %al
  203cfd:	c3                   	ret
```
vs the uglier code
```asm
0000000000203d50 <is_json_whitespace__T_u8>:
  203d50:	40 80 ff 21          	cmp    $0x21,%dil
  203d54:	0f 92 c1             	setb   %cl
  203d57:	48 b8 00 26 00 00 01 	movabs $0x100002600,%rax
  203d5e:	00 00 00 
  203d61:	48 0f a3 f8          	bt     %rdi,%rax
  203d65:	0f 92 c0             	setb   %al
  203d68:	20 c8                	and    %cl,%al
  203d6a:	c3                   	ret
```

I'll open an issue to follow up on this

Reply:
That's a shame... But also wow, that ugly code optimizes amazingly well, llvm is just showing off.

Reply:
Yep, very impressive!